python/ray/serve/router.py

```diff
@@ -55,148 +57,158 @@ def __init__(
                 (LongPollNamespace.REPLICA_HANDLES, backend_tag): self.
                 update_worker_replicas,
-            },
-            call_in_event_loop=event_loop,
-        )
+            })
 
     def set_max_concurrent_queries(self, backend_config: BackendConfig):
         new_value: int = backend_config.max_concurrent_queries
         if new_value != self.max_concurrent_queries:
             self.max_concurrent_queries = new_value
             logger.debug(
                 f"ReplicaSet: changing max_concurrent_queries to {new_value}")
-            self.config_updated_event.set()
+        self._drain_pending_queue()
 
     def update_worker_replicas(self, worker_replicas: Iterable[ActorHandle]):
-        added, removed, _ = compute_iterable_delta(
-            self.in_flight_queries.keys(), worker_replicas)
+        with self.lock:
+            added, removed, _ = compute_iterable_delta(
+                self.in_flight_queries.keys(), worker_replicas)
 
-        for new_replica_handle in added:
-            self.in_flight_queries[new_replica_handle] = set()
+            for new_replica_handle in added:
+                self.in_flight_queries[new_replica_handle] = set()
 
-        for removed_replica_handle in removed:
-            # Delete it directly because shutdown is processed by controller.
-            del self.in_flight_queries[removed_replica_handle]
+            for removed_replica_handle in removed:
+                # Delete it directly because shutdown is processed
+                # by controller.
+                del self.in_flight_queries[removed_replica_handle]
 
-        if len(added) > 0 or len(removed) > 0:
-            self.replica_iterator = itertools.cycle(
-                self.in_flight_queries.keys())
-            logger.debug(
-                f"ReplicaSet: +{len(added)}, -{len(removed)} replicas.")
-            self.config_updated_event.set()
+            if len(added) > 0 or len(removed) > 0:
+                self.replica_iterator = itertools.cycle(
+                    self.in_flight_queries.keys())
+                logger.debug(
+                    f"ReplicaSet: +{len(added)}, -{len(removed)} replicas.")
+                self._drain_pending_queue()
 
-    def _try_assign_replica(self, query: Query) -> Optional[ray.ObjectRef]:
-        """Try to assign query to a replica, return the object ref if succeeded
-        or return None if it can't assign this query to any replicas.
+    def _on_query_tracker_completed(
+            self,
+            _trakcer_result: Any,
+            *,
+            replica_handle: ActorHandle,
+            tracker_ref: ray.ObjectRef,
+    ):
+        """Callback after each query is completed."""
+        with self.lock:
+            if replica_handle in self.in_flight_queries:
+                self.in_flight_queries[replica_handle].remove(tracker_ref)
+                self._drain_pending_queue()
+
+    def _drain_pending_queue(self):
```

Comment:
Oh, I think this method is sort of mis-named. This is really just the normal codepath for sending queries, right? I would call it something like `try_send_pending_queries`. This name made me assume that we would send queries in a loop until the pending queue was completely drained.